File: Day04/d4.py
with open("input.txt", "r") as inputFile:
    input = [list(map(int, x.replace('-', ',').split(','))) for x in inputFile.read().splitlines()]
    count = count2 = 0
    for a, b, x, y in input:
        if (a >= x and b <= y) or (x >= a and y <= b):
            count += 1
            count2 += 1
        elif (a <= x <= b) or (a <= y <= b):
            count2 += 1
    print(count)
    print(count2)

File: Day04/d4.hs
parseInput:: [[Char]] -> [[Int]]
parseInput = map(\y -> read ("[" ++ [if x == '-' then ',' else x | x <- y] ++ "]")) 

main = do
    input <- parseInput . lines <$> readFile "input.txt"
    print $ part1 input
    print $ part2 input

part1 input = sum $ fullyContains <$> input
part2 input = sum $ overlap <$> input

fullyContains [a, b, x, y] 
  | (a >= x && b <= y) || (x >= a && y <= b) = 1
  | otherwise = 0

overlap [a, b, x, y] 
  | (a <= x && x <= b) || (a <= y && y <= b) = 1
  | otherwise = fullyContains [a, b, x, y]
